### PR TITLE
Correct implementation of Facebook-OAuth

### LINKED ISSUE #2298: Facebook Login doesn't work
**Describe the bug**
Facebook is implemented using Torii which is not integrating properly with the backend. 

**Expected behavior**
Integrate Facebook login with frontend properly and delete the repeated functionalities.

**Additional context**
Working on it.

### PR DESCRIPTION
**Pull request by prateekj117**
#### Checklist

- [X] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [X] My branch is up-to-date with the Upstream `development` branch.
- [ ] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:

- Use of torii is not required and is removed
- Integrate Facebook OAuth functionality with server

Fixes #2298

### CODE REVIEW

**Comment by CosmicCoder96** [comment]
I am not sure what this is, if you don't persist the session, how will the auth manager know, either way please do not open prs on the frontend for functionality which is not merged on the server.

**Comment by prateekj117** [comment]
@CosmicCoder96 After a very long discussion with @iamareebjamal , we decided to take torii out of play on frontend. The whole implementation of facebook oauth was wrong, and there was no communication properly between backend and frontend. So, we decided to reimplement it all, and take torii out of play.

**Comment by iamareebjamal** [comment]
@CosmicCoder96 We need these changes in order for OAuth to work. Server changes are required as well. This is a WIP PR. If we close this until PR on server is merged, we'll be locked in a cycle. I request you to review it once it is complete. You can read the relevant discussion about changes required in OAuth system on Gitter chat on open event server room

**Comment by prateekj117** [comment]
@iamareebjamal Please review.

**Comment by prateekj117** [comment]
@CosmicCoder96 Please review now.

File: app/components/forms/login-form.js [hunk 1]
```diff
@@ -84,27 +84,10 @@ export default Component.extend(FormMixin, {
...
-                      this.get('authManager').persistCurrentUser(
-                        await this.get('store').findRecord('user', tokenPayload.identity)
-                      );
-                    }
-                  });
-              });
-          });
+          this.get('loader').load('/auth/oauth/facebook')
+            .then(async response => {
+              window.location.replace(response.url);
```
Comment (iamareebjamal) [inline app/components/forms/login-form.js]: Should open a popup instead

File: app/routes/facebook.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import Route from '@ember/routing/route';
+
+export default Route.extend({
+  queryParams: {
+    code: ''
+  },
+  model(params) {
+    this.get('loader').post(`/auth/oauth/login/facebook?code=${ params.code }`)
```
Comment (iamareebjamal) [inline app/routes/facebook.js]: Should use the provider query param instead

File: app/routes/facebook.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      .then(response => {
+        let credentials = {
+              identification : response.email,
+              password       : response.facebook_login_hash
```
Comment (iamareebjamal) [inline app/routes/facebook.js]: Change server response from facebook_login_hash to oauth_hash

File: app/routes/facebook.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import Route from '@ember/routing/route';
+
+export default Route.extend({
+  queryParams: {
+    code: ''
+  },
+  model(queryParams) {
+    this.get('loader').post(`/auth/oauth/login/facebook?code=${ queryParams.code }`)
```
Comment (iamareebjamal) [inline app/routes/facebook.js]: Don't hardcode 'facebook'. Use the provider query parameter

**Comment by iamareebjamal** [comment]
@CosmicCoder96 Can you please review this and merge. Need this to go ahead with other OAuth implementations

File: app/routes/facebook.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import Route from '@ember/routing/route';
+
+export default Route.extend({
+  queryParams: {
+    code     : '',
+    provider : ''
+  },
+  model(queryParams) {
+    this.get('loader').post(`/auth/oauth/login/${ queryParams.provider }?code=${ queryParams.code }`)
```
Comment (CosmicCoder96) [inline app/routes/facebook.js]: All this controlling logic should be inside the controller of the route. Not the model hook.

File: app/routes/facebook.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        this.get('session')
+          .authenticate(authenticator, credentials)
+          .then(async() => {
+            const tokenPayload = this.get('authManager').getTokenPayload();
+            if (tokenPayload) {
+              this.get('authManager').persistCurrentUser(
+                await this.get('store').findRecord('user', tokenPayload.identity)
+              );
+            }
+            window.location.replace('/');
```
Comment (CosmicCoder96) [inline app/routes/facebook.js]: That is not how you transition in an ember app. Use the native methods of ‘transitionTo’ and ‘transitionToRoute’ instead.

File: app/routes/facebook.js (hunk 5 above)
Comment (prateekj117) [reply app/routes/facebook.js]: @CosmicCoder96 @niranjan94 Need some help here. I tried making a controller for the route 'oauth/callback' and a js file in controller folder with the same path, and noticed it doesn't get called when the route is accessed.

Comment (prateekj117) [reply app/routes/facebook.js]: Screenshot from 2019-03-23 00-00-02
Here is the file path.

Comment (CosmicCoder96) [reply app/routes/facebook.js]: if your route name is facebook then the controller should also be at routes/facebook.

Comment (prateekj117) [reply app/routes/facebook.js]: @CosmicCoder96 In routes/facebook or controller/facebook?

Comment (CosmicCoder96) [reply app/routes/facebook.js]: `controller/facebook`

**Comment by prateekj117** [comment]
@CosmicCoder96 @iamareebjamal Please review now.

File: app/components/forms/login-form.js (hunk 1 above)
Comment (CosmicCoder96) [inline app/components/forms/login-form.js]: use ember methods.

File: app/components/forms/login-form.js [hunk 7]
```diff
@@ -84,27 +84,10 @@ export default Component.extend(FormMixin, {
     async auth(provider) {
       try {
         if (provider === 'facebook') {
-          this.get('torii').open('facebook').then(authData => {
```
Comment (CosmicCoder96) [inline app/components/forms/login-form.js]: remove torii completely, from the app. Not just here. (uninstall it)

File: app/components/forms/login-form.js (hunk 1 above)
Comment (prateekj117) [reply app/components/forms/login-form.js]: @CosmicCoder96 This being an external url cannot be redirected using transitionTo.

Comment (prateekj117) [reply app/components/forms/login-form.js]: @CosmicCoder96 If there is any other method for this, can you please guide me through it.

Comment (prateekj117) [reply app/components/forms/login-form.js]: @CosmicCoder96 Also, stripe is using torii, and I am not sure if it is required there or not, so I don't think we should uninstall it completely. I have deleted the torii functionality for facebook though. So can you please review the changes again.

**Pull request merged**